libs/src/services/solanaWeb3Manager/rewards.js

```diff
@@ -0,0 +1,767 @@
+  /**
+   * Creates an instance of SubmitAttestationInstructionData.
+   * @param {{transferId: string}} {
+   *     transferId
+   *   }
+   * @memberof SubmitAttestationInstructionData
+   */
+  constructor({ transferId }) {
+    this.id = transferId
+  }
+}
+
+const submitAttestationInstructionSchema = new Map([
+  [
+    SubmitAttestationInstructionData,
+    {
+      kind: "struct",
+      fields: [["id", "string"]],
+    },
+  ],
+])
+
+class ValidateAttestationsInstructionData {
+  /**
+   * Creates an instance of ValidateAttestationsInstructionData.
+   * @param {{
+   *     amount: number,
+   *     id: string,
+   *     eth_recipient: Uint8Array
+   *   }} {
+   *     amount,
+   *     id,
+   *     eth_recipient
+   *   }
+   * @memberof ValidateAttestationsInstructionData
+   */
+  constructor({ amount, id, eth_recipient }) {
+    this.amount = amount
+    this.id = id
+    this.eth_recipient = eth_recipient
+  }
+}
+
+const validateAttestationsInstructionSchema = new Map([
+  [
+    ValidateAttestationsInstructionData,
+    {
+      kind: "struct",
+      fields: [
+        ["amount", "u64"],
+        ["id", "string"],
+        ["eth_recipient", [20]],
+      ],
+    },
+  ],
+])
+
+/**
+ * @typedef {Object} AttestationMeta
+ * @property {string} ethAddress
+ * @property {string} signature
+ */
+
+/**
+ * Submits attestations from Discovery Nodes and AAO that a user has completed a challenge.
+ *
+ * @param {{
+ *   rewardManagerProgramId: PublicKey,
+ *   rewardManagerAccount: PublicKey,
+ *   attestations: AttestationMeta[]
+ *   oracleAttestation: AttestationMeta
+ *   challengeId: string,
+ *   specifier: string,
+ *   feePayer: PublicKey,
+ *   attestationSignature: string,
+ *   recipientEthAddress: string,
+ *   tokenAmount: BN
+ *   identityService: any
+ *   connection: Connection
+ * }} {
+ *   rewardManagerProgramId,
+ *   rewardManagerAccount,
+ *   attestations,
+ *   oracleAttestation,
+ *   challengeId,
+ *   specifier,
+ *   feePayer,
+ *   recipientEthAddress,
+ *   tokenAmount,
+ *   identityService,
+ *   connection
+ * }
+ */
+export async function submitAttestations({
+  rewardManagerProgramId,
+  rewardManagerAccount,
+  attestations,
+  oracleAttestation,
+  challengeId,
+  specifier,
+  feePayer,
+  recipientEthAddress,
+  tokenAmount,
+  identityService,
+  connection,
+}) {
+  // Construct combined transfer ID
+  const transferId = constructTransferId(challengeId, specifier)
+
+  // Derive the message account we'll use to store the attestations
+  const [
+    rewardManagerAuthority,
+    derivedMessageAccount,
+  ] = await deriveMessageAccount(
+    transferId,
+    rewardManagerProgramId,
+    rewardManagerAccount
+  )
+
+  // Add instructions from DN attestations - each attestation
+  // needs a pairing of SECP recovery instruction and submit
+  // attestation instruction.
+  let instructions = await Promise.all(
+    attestations.reduce((instructions, meta, i) => {
```

Comment:
how come not a reduce(async () => {....})

Reply:
Couldn't get this way to work for some reason. Spent a while trying it